### ISSUE TITLE
Kubernetes client request times out when called from a vert.x eventbus consumer

### ISSUE THREAD

**Issue by thomascube** [quarkusio/quarkus#38133]
### Describe the bug

In our application we have a Kubernetes informer listening on ConfigMaps and posting changes to the Vert.x event bus. A consumer listening on these eventbus messages will then fetch another resource from Kubernetes using the same client. These requests do not complete and eventually are aborted with a timeout exception.

### Expected behavior

The [KubernetesClient](https://quarkus.io/guides/kubernetes-client) can be used in all contexts throughout the application and does not timeout.

### Actual behavior

The same request to the Kubernetes API with the same KubernetesClient succeeds from another context (e.g. a REST Call) but blocks and fails when called from an eventbus consumer method.

This is the exception and stack trace we see when calling the Kubernetes API from the eventbus consumer:
```posh
 io.fabric8.kubernetes.client.KubernetesClientException
        at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:509)
        at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:524)
        at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleGet(OperationSupport.java:467)
        at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.handleGet(BaseOperation.java:791)
        at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.requireFromServer(BaseOperation.java:192)
        at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.get(BaseOperation.java:148)
        at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.get(BaseOperation.java:97)
        at ch.sbb.esta.service.KubernetesApiService.getConfigMapByName(KubernetesApiService.java:16)
        at ch.sbb.esta.service.KubernetesApiService_ClientProxy.getConfigMapByName(Unknown Source)
        at ch.sbb.esta.service.ConfigMapListener.configMapModified(ConfigMapListener.java:26)
        at ch.sbb.esta.service.ConfigMapListener_ClientProxy.configMapModified(Unknown Source)
        at ch.sbb.esta.service.ConfigMapListener_VertxInvoker_configMapModified_7811ecff56b0d4c0b09e6cf8d655b75eea64c1e9.invokeBean(Unknown Source)
        at io.quarkus.vertx.runtime.EventConsumerInvoker.invoke(EventConsumerInvoker.java:45)
        at io.quarkus.vertx.runtime.VertxEventBusConsumerRecorder$3$1.handle(VertxEventBusConsumerRecorder.java:167)
        at io.quarkus.vertx.runtime.VertxEventBusConsumerRecorder$3$1.handle(VertxEventBusConsumerRecorder.java:111)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:277)
        at io.vertx.core.eventbus.impl.MessageConsumerImpl.dispatch(MessageConsumerImpl.java:177)
        at io.vertx.core.eventbus.impl.HandlerRegistration$InboundDeliveryContext.execute(HandlerRegistration.java:137)
        at io.vertx.core.eventbus.impl.DeliveryContextBase.next(DeliveryContextBase.java:80)
        at io.vertx.core.eventbus.impl.DeliveryContextBase.dispatch(DeliveryContextBase.java:43)
        at io.vertx.core.eventbus.impl.HandlerRegistration.dispatch(HandlerRegistration.java:98)
        at io.vertx.core.eventbus.impl.MessageConsumerImpl.deliver(MessageConsumerImpl.java:183)
        at io.vertx.core.eventbus.impl.MessageConsumerImpl.doReceive(MessageConsumerImpl.java:168)
        at io.vertx.core.eventbus.impl.HandlerRegistration.lambda$receive$0(HandlerRegistration.java:49)
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.util.concurrent.TimeoutException
        at io.fabric8.kubernetes.client.utils.AsyncUtils.lambda$withTimeout$0(AsyncUtils.java:42)
        at io.fabric8.kubernetes.client.utils.Utils.lambda$schedule$6(Utils.java:473)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        ... 1 more

```

### How to Reproduce?

See this minimal example project: https://github.com/thomascube/quarkus-kubernetes-informer
Steps to reproduce are described in the README there.

### Output of `uname -a` or `ver`

Darwin Kernel Version 23.1.0

### Output of `java -version`

OpenJDK Runtime Environment Temurin-21.0.1+12 (build 21.0.1+12-LTS)

### Quarkus version or git rev

3.6.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.6

### Additional information

If this is not an actual bug, I'm glad the get suggestions how to configure the KubernetesClient to work in all contexts of a Quarkus application. I already tried to create different instances of a KubernetesClient but that didn't resolve the problematic behavior.

**Comment by geoand** [comment]
This souns like a bug in the Kubernetes Client - cc @manusa

**Comment by manusa** [comment]
Might have something to do with the Vertx thread pool and running blocking code (such as the kubernetes request) in that method.

Have you tried using `@ConsumeEvent(value = "CM_MODIFIED", blocking = true)` instead?

**Comment by thomascube** [comment]
> Have you tried using `@ConsumeEvent(value = "CM_MODIFIED", blocking = true)` instead?

Indeed, adding`blocking = true` seems to solve the problem. Thanks a lot!

I leave it up to you, whether you want to keep the issue open for further investigation or close it.

**Comment by manusa** [comment]
> I leave it up to you, whether you want to keep the issue open for further investigation or close it.

I think that it might be a documentation problem, not a bug @geoand?

AFAIK you should avoid doing time-consuming tasks (or any blocking call at all) within a `@ConsumeEvent` method unless you use the `blocking` option or you're wrapping it within a `vertx.executeBlocking` lambda.

**Comment by geoand** [comment]
We can definitely do with an improvement in the documentation, but I am also thinking we might be able to do better...

Essentially what happens here is the following:

* Execution is performed on an event loop thread
* The Kubernetes Client makes a call to the API server
* As Vert.x is being used under the hood, the IO request to the API server is done on the same event loop thread, but because the client needs to present a blocking API, it has to wait for the result 
  * This waiting essentially stals all progress on that event loop thread since the thread that is supposed to be handling IO events is blocked waiting on a result that would under normal circumestances show up at some point on it's processing queue.

What I propose we do (is what we have already done in other places), which is to throw an actionable exception when the client attempts to wait on an event loop thread.

As far as I can tell, the wait occurs in `OperationSupport.waitForResult(OperationSupport.java:509)`
so the client provides some kind of SPI, Quarkus could hook into the necessary check and throw the proper exception if needed.

**Comment by manusa** [comment]
> What I propose we do (is what we have already done in other places), which is to throw an actionable exception when the client attempts to wait on an event loop thread.

This seems like a good UX improvement.

> Quarkus could hook into the necessary check and throw the proper exception if needed.

This is sort of HTTPClient implementation-wise. If a user is using a different HTTP client (to Vert.x), then this wouldn't apply.

I'm also thinking that his might be trouble if the Kubernetes Client is with other libraries besides from Quarkus.

I have no clue on how to do this at the moment, but my understanding is that if we could implement the detection mechanism in the httpclient-vertx module we would be fixing be providing a consistent experience for any kind of downstream usage.

@shawkins, any thoughts?

**Comment by geoand** [comment]
> This is sort of HTTPClient implementation-wise. If a user is using a different HTTP client (to Vert.x), then this wouldn't apply.

Sure yeah - it's up to you what the proper abstraction would be that we could use on the Quarkus side.

**Comment by shawkins** [comment]
> As far as I can tell, the wait occurs in OperationSupport.waitForResult(OperationSupport.java:509)
so the client provides some kind of SPI, Quarkus could hook into the necessary check and throw the proper exception if needed.

That is the main wait for the blocking rest calls. It unfortunately isn't the only wait - there's a blocking way to start the leader elector, waitUntilReady/Condition calls, and some smaller (and probably more waits) that exist in interal logic like behind createOrReplace. Since how the wait is enforced may not be tied directly to the completablefuture returned by the vertx http client it seems like we'd have to update any and all of these locations to make a call into the spi.

**Comment by herasimau** [comment]
Started facing this issue after updating from Quarkus 2.16.5 to Quarkus 3.6.5

**Comment by manuelwallrapp** [comment]
Have the same issue here when I chain calls in Quarkus in this way:
1. REST Service gets called on my Quarkus App
2. If I have in the call chain something likes this: 
3. return myPipelineRunUni.map(pr -> makeAHttpCallWithVertxWebClient(pr)).map(p->makeACallToKubernetes(pr)). This call will then Timeout in the Kubernetes Api. If I don't do the makeAHttpCallWithVertxWebClient, the call  makeACallToKubernetes will not Timeout. So there seems to be a problem with the thread handling.

**Comment by thomascube** [comment]
Dear all

In order to demonstrate the case @manuelwallrapp describes above, I have extend the test case to isolate the issue. 
The timeout also happens when using the KubernetesClient in an asynchronous chain after performing a HTTP request using the Vert.x WebClient. See https://github.com/thomascube/quarkus-kubernetes-informer/blob/master/src/main/java/ch/sbb/esta/ConfigMapResource.java#L37-L42

**Comment by manuelwallrapp** [comment]
Will anybody take care of this issue? Would be great if somebody could fix this.

**Comment by cescoffier** [comment]
First, it's the expected behavior - you need to mark your consumer as blocking if you are doing blocking calls.  Otherwise, you need what @thomascube has implemented (an asynchronous way to do the same thing)

I would close it as it's not a bug. We can improve the reporting, but that would be done in a separate task (and I'm not sure it should be done in Quarkus core)
